src/test/tools/ddmap/ddmap.c

```diff
@@ -202,21 +208,43 @@ ddmap_write_data(const char *path, const char *data,
  */
 static int
 ddmap_write_from_file(const char *path_in, const char *path_out,
-	os_off_t offset_in, os_off_t offset_out, size_t len)
+	os_off_t offset_in, os_off_t offset_out, unsigned long bytes,
+	unsigned long count)
 {
-	char *src;
+	char *src, *tmp_src;
 	ssize_t file_in_size = util_file_get_size(path_in);
-
-	if ((size_t)file_in_size < len + (size_t)offset_in) {
-		outv_err("offset with length exceed input file size");
-		return -1;
-	}
+	size_t data_left, len;
 
 	util_init();
 	src = util_file_map_whole(path_in);
 	src += offset_in;
 
-	ddmap_write_data(path_out, src, offset_out, len);
+	if (!count)
+		count = 1;
+
+	data_left = (size_t)file_in_size;
+	tmp_src = src;
+	do {
+		if (data_left >= (size_t)bytes)
+			len = (size_t)bytes;
+		else
+			len = data_left;
+
+		ddmap_write_data(path_out, tmp_src, offset_out, len);
+		data_left = MAX(0, data_left - (size_t)bytes);
+
+		if (!data_left) {
+			data_left = (size_t)file_in_size;
+			tmp_src = src;
+		} else {
+			data_left -= len;
+			tmp_src += (os_off_t)len;
+		}
```

Comment:
This is really odd code. If I understand this code correctly you could replace the part after ddmap_write_data with:

```
tmp_src += len;
data_left -= len;
if (data_left == 0) {
   /* copy from the beginning */
   data_left = file_in_size;
   tmp_src = src;
}
```